typed-protocols/src/Network/TypedProtocol/Driver.hs

```diff
@@ -170,47 +197,83 @@ runPipelinedPeer maxOutstanding tr codec channel (PeerPipelined peer) = do
-    go n    (SenderEffect k) = k >>= go n
-    go Zero (SenderDone _ x) = return x
+    go :: forall st' n.
+          Nat n
+       -> Maybe bytes
+       -> PeerSender ps pr st' n c m a
+       -> m a
+    go n     trailing (SenderEffect k) = k >>= go n trailing
+    go Zero _trailing (SenderDone _ x) = return x
+    --TODO: same issue with trailing as the 'Done' case for 'runPeer' above.
 
-    go n (SenderYield stok msg k) = do
+    go Zero Nothing (SenderYield stok msg k) = do
       traceWith tr (TraceSendMsg (AnyMessage msg))
       send (encode stok msg)
-      go n k
+      go Zero Nothing k
+
+    go Zero trailing (SenderAwait stok k) = do
+      decoder <- decode stok
+      res <- runDecoderWithChannel channel trailing decoder
+      case res of
+        Right (SomeMessage msg, trailing') -> do
+          traceWith tr (TraceRecvMsg (AnyMessage msg))
+          go Zero trailing' (k msg)
+        Left failure ->
+          throwM failure
 
-    go n (SenderPipeline stok msg receiver k) = do
-      atomically (writeTBQueue receiveQueue (ReceiveHandler receiver))
+    go n Nothing (SenderPipeline stok msg receiver k) = do
+      atomically (writeTBQueue receiveQueue (ReceiveHandler n receiver))
       traceWith tr (TraceSendMsg (AnyMessage msg))
       send (encode stok msg)
-      go (Succ n) k
+      go (Succ n) Nothing k
 
-    go (Succ n) (SenderCollect Nothing k) = do
+    go (Succ n) Nothing (SenderCollect Nothing k) = do
       c <- atomically (readTBQueue collectQueue)
-      go n (k c)
+      go n Nothing (k c)
 
-    go (Succ n) (SenderCollect (Just k') k) = do
+    go (Succ n) Nothing (SenderCollect (Just k') k) = do
       mc <- atomically (tryReadTBQueue collectQueue)
       case mc of
-        Nothing -> go (Succ n) k'
-        Just c  -> go n (k c)
+        Nothing -> go (Succ n) Nothing  k'
+        Just c  -> go n        Nothing (k c)
+
+    -- These cases are protocol errors. The only times we can have trailing
+    -- data on the sending side thread is after a /non-pipelined/ Await when
+    -- we're still in a state where the remote peer has agency. Why? Because
+    -- those are the cases when you can have two incoming non-pipelined
+    -- messages back to back, which is the only time you can have trailing
+    -- data on the sending side thread.
+    --
+    -- We need another argument to explain why handling trailing data in the
+    -- sending and receiving threads can be independent rather than having
+    -- to be synchronised. The argument is that when there's no outstanding
+    -- pipelined messages then the trailing data on the receiving side must be
+    -- empty and thus it is ok to handle
+    -- trailing data on the send side if we now do synchronous non-pipelined
+    -- yield and multiple awaits.
+    --
+    go _ Just{} SenderYield{}    = throwM IllegalTrailingData
+    go _ Just{} SenderPipeline{} = throwM IllegalTrailingData
+    go _ Just{} SenderCollect{}  = throwM IllegalTrailingData
```

Comment:
It seems we only need to `throwM` in the `SenderCollect` case, but you're right that all three cases are protocol errors.

Reply:
It's not so clear to me whether they are, though. Imagine a protocol where agent `A` has a choice of 2 or more transitions, such that no matter which transition is chosen, `B` has agency next, and there is precisely one transition that `B` can make: all of those transitions from `A` land in the same state, and that state has only one outgoing transition.

Then is it a protocol error for `B` to send its transition _before_ it receives `A`'s choice?

Reply:
So it'd involve two pipelined peers talking to each other. And in the case that peer A is doing a synchronous await, peer B taking advantage of a protocol where peer A can only transition to one state where B has agency and so B can send its response before it sees the request.

Reply:
To cope, I think we'd need to send the trailing data over to the next pipelined response handler.

The worse case is the other way, if you want to allow trailing data that arrives before we do a pipelined request. Again, that'd rely on making a response before the request is known.

Reply:
> So perhaps the more accurate description of the above is that this pipelined driver cannot cope with such situations, and this check ensures it detects and fails in such situations.

Why can't it cope? When it gets to reading on the next await, it will use the trailing bytes to decode the valid message.

Reply:
> So yes this is possible, but it relies on two pipelined peers and a protocol of a particular shape and one where the response does not depend on the request.

Yes. And the response not depending on the request is the necessary condition for pipelining. It's just to say: it depends upon the protocol having a form which admits pipelining.

My issue here is with the scope of typed-protocols. It shouldn't care, nor attempt to determine, _when_ the other peer sent the message. All it should be concerned with is whether that message is of the expected type (as determined by the protocol transitions).

Reply:
> Why can't it cope? When it gets to reading on the next await, it will use the trailing bytes to decode the valid message.

Because the implementation does not synchronise the trailing data it gets on the "sender" thread with the trailing data on the "receiver" thread. I was then going to rely on not having to handle these cases to avoid having to synchronise.

Let me think if it's not too tricky to do that synchronisation. Perhaps it's not because we make sure not to mix the pipelined and syncronous cases, so perhaps we can simply pass the trailing back and forth at the transition boundaries.

Reply:
> Yes. And the response not depending on the request is the necessary condition for pipelining. It's just to say: it depends upon the protocol having a form which admits pipelining.

It's a bit weirder than that. All the normal cases are one peer is pipelined and the other is not, and responses have a data dependency on requests. This would require both peers to be pipelined and the response to be sent before the request, so have no data dependency.

The normal thing is that my next pipelined request does not depend on the result of my previous pipelined request.

Reply:
> Let me think if it's not too tricky to do that synchronisation. Perhaps it's not because we make sure not to mix the pipelined and syncronous cases, so perhaps we can simply pass the trailing back and forth at the transition boundaries.

Because we run two threads: sender and receiver we need to pass the trailing bytes in a mutable cell, and this is exactly what I did in the `coot/sender-await` branch.

Reply:
Maybe it's also possible to start a listening thread when we start pipelining, and then await on it when all all the results are collected (the driver knows when that happens: `go Zero (SenderAwait _ _)`).  This way we could pass the trailing bytes between the two threads.  The listening thread life span is embedded only during pipelining.

Reply:
Would it be feasible to drive a pipelined peer without multithreading? What if we had `peek :: Channel m t -> m (Maybe t)` (implemented using non-blocking sockets in case of TCP and similar)?

For the `SenderCollect (Just k') k` case, we put any available bytes from `peek` through the incremental decoder and, if it's finished, use `k'`, otherwise, use `k` with the new incremental decoder continuation.

This way, no bytes are cleared from the channel until a `SenderCollect` or `SenderAwait` is reached, but that should be fine. Something the programmer would expect as well (backpressure).

Reply:
We should try, it seems possible to do that.